### PR TITLE
Consumer resource labels

### PR DESCRIPTION
**Pull request by philipnee**
https://confluentinc.atlassian.net/browse/CLIENTS-1854

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java [hunk 1]
```diff
@@ -385,4 +403,45 @@ public static PushTelemetryRequest.Builder createPushTelemetryRequest(boolean te
             metricsData);
     }
 
+    public static ResourceLabel createConsumerLabels(final ConsumerGroupMetadata groupMetadata) {
+        if(groupMetadata == null) {
+            return new ResourceLabel();
+        }
+
+        ResourceLabel resourceLabels = new ResourceLabel();
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java:411]: Do you know if the group metadata is available at the time of construction? Or is it something that has to be retrieved from the cluster on first contact?

File: clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java [hunk 2]
```diff
@@ -385,4 +403,45 @@ public static PushTelemetryRequest.Builder createPushTelemetryRequest(boolean te
...
+        // TODO: constant-lize these keys
+        resourceLabels.maybePut("group.member.id", groupMetadata.memberId());
+        resourceLabels.maybePut("group.id", groupMetadata.groupId());
+        resourceLabels.maybePut("group.instance.id", groupMetadata.groupInstanceId().get());
+
+        return resourceLabels;
+    }
+
+    static class ResourceLabel {
+        private Map<String, String> resourceLables;
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java:421]:
```suggestion
        private Map<String, String> resourceLabels;
```

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java:421]: There's a typo in the word "labels."

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 3]
```diff
@@ -363,7 +363,7 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
                     config.originalsWithPrefix(CommonClientConfigs.METRICS_CONTEXT_PREFIX));
             this.metrics = new Metrics(metricConfig, reporters, time, metricsContext);
             this.producerMetrics = new KafkaProducerMetrics(metrics);
-            this.clientTelemetry = ClientTelemetryUtils.create(config, time, clientId);
+            this.clientTelemetry = ClientTelemetryUtils.create(config, time, clientId, Collections.singletonMap("transactional.id", transactionalId));
```
[the remark this answers is not in the document]
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:366]: I guess `maybePut` handles filtering this out deeper down, right?

**Review by kirktrue — CHANGES_REQUESTED** [review]
Looks good overall.

I don't think it will pass the linter, though, due to some missing spaces around brackets and parenthesis.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:366]: i say let's omit it, i think it only exists in the transactional producer, so we don't need to log that in the idempotent prod.

File: clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java [hunk 4]
```diff
@@ -233,16 +239,28 @@ public static ByteBuffer serialize(Collection<TelemetryMetric> telemetryMetrics,
...
+
+    public static ClientTelemetry create(AbstractConfig config, Time time, String clientId, ResourceLabel labels) {
         if (config == null)
             throw new IllegalArgumentException("config for ClientTelemetry cannot be null");
 
         boolean enableMetricsPush = config.getBoolean(CommonClientConfigs.ENABLE_METRICS_PUSH_CONFIG);
-        return create(enableMetricsPush, time, clientId);
+
+        if(!config.getString(ConsumerConfig.CLIENT_RACK_CONFIG).isEmpty())
+            labels.maybePut("client.rack", config.getString(ConsumerConfig.CLIENT_RACK_CONFIG));
```
[the remark this answers is not in the document]
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java:256]: ah... i didn't notice it was Consumer specific config, though, i think we want client rack for all clinets, let me look into that.

**Comment by philipnee** [comment]
@kirktrue - Thanks for the reviews! Back at you!

File: clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java (hunk 1 above)
Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/telemetry/ClientTelemetryUtils.java]: Actually - good point, I think it is available via the coordinator, which is initialized after the client telemetry. I'll look into it

**Pull request closed without merging**